Go: Make flow configurations use new data flow API

owen-mc opened the pull request:
I highly recommend reviewing commit by commit (and spreading it out over several sessions). All configurations have tests (normally of the related query). There are no changes in results, though sometimes there are changes in exact `PathNode`s and edges between them.

owen-mc commented:
I will rebase on main when https://github.com/github/codeql/pull/13822 is merged, which will fix the sync files check.

mbg commented on go/ql/lib/semmle/go/security/IncorrectIntegerConversionLib.qll:
```diff
@@ -148,6 +150,115 @@ class ConversionWithoutBoundsCheckConfig extends TaintTracking::Configuration {
 }
 
+/** Flow state for ConversionWithoutBoundsCheckConfig. */
+newtype MyFlowState =
```
Should this have a less generic name, e.g. `IntegerConversionState`?

mbg commented on go/ql/test/library-tests/semmle/go/frameworks/Spew/TaintFlows.ql:
```diff
@@ -1,28 +1,18 @@
-  override predicate isSource(DataFlow::Node source) { source instanceof UntrustedFlowSource }
+module TestConfig implements DataFlow::ConfigSig {
+  predicate isSource(DataFlow::Node source) {
+    source =
+      any(Function f | f.getName() = ["getUntrustedString", "getUntrustedStruct"])
+          .getACall()
+          .getResult()
```
I note that the test previously ranged over all of `UntrustedFlowSource` but now is an inlined `UntrustedSource`. Do we miss out anything important by not having this range over all of `UntrustedFlowSource` anymore?

owen-mc replied:
I don't think so - we're just trying to match the two functions used as sources in the test files. I can only imagine was written like this to be more like a real source definition.

mbg commented on go/ql/test/library-tests/semmle/go/frameworks/Protobuf/TaintFlows.ql:
```diff
@@ -1,27 +1,18 @@
-  override predicate isSource(DataFlow::Node source) { source instanceof UntrustedFlowSource }
+module Config implements DataFlow::ConfigSig {
+  predicate isSource(DataFlow::Node source) {
+    exists(Function fn | fn.hasQualifiedName(_, ["getUntrustedString", "getUntrustedBytes"]) |
+      source = fn.getACall().getResult()
+    )
```
This previously ranged over all of `UntrustedFlowSource`, but is now more limited. Do we miss out on anything as a result of this change?

owen-mc replied:
As above, I don't think it matters for tests.

mbg commented on go/ql/test/library-tests/semmle/go/frameworks/K8sIoApiCoreV1/main.go:
```diff
@@ -9,86 +9,88 @@ import (
 
 func source() interface{} {
-	return make([]byte, 1, 1)
+	return make([]byte, 1)
```
What's the reason for this change?

owen-mc replied:
My IDE told me it was better to do it this way. Presumably it's a built-in Go linter of some kind.

mbg commented on go/ql/test/library-tests/semmle/go/frameworks/EvanphxJsonPatch/TaintFlows.ql:
```diff
@@ -1,33 +1,20 @@
     )
   }
+
+  int fieldFlowBranchLimit() { result = 1000 }
```
Could you add a comment noting why you added this here for this particular configuration?

owen-mc replied:
It isn't needed. I've removed it. I must have added it by mistake, probably copying from somewhere else.

mbg reviewed (commented):
Good work updating all of this! ✨ The refactoring generally looks reasonable. There are a few specific comments I had while looking over this and two more broad ones:

- For queries where the expected test output changed and you haven't done so yet, it would be good to note why it changed in the commit description or in the PR description.
- Most/all of the old data/taint flow configurations had docs comments explaining their purpose. However, the new `ConfigSig` implementations and data/taint flow configurations mostly do not have matching docs comments. In general, it would be nice to have more documentation/comments.

There also seem to a good few `ConfigSig`s which are basically the same as a result of the `[..]Customizations` approach along the lines of:
```ql
private module Config implements DataFlow::ConfigSig {
    predicate isSource(DataFlow::Node source) { source instanceof Source }

    predicate isSink(DataFlow::Node sink) { sink instanceof Sink }

    predicate isBarrier(DataFlow::Node node) { node instanceof Barrier }
}
```
Would it be possible (in the future, not as part of this PR) to refactor this so that we have a generic implementation that we can just parameterise over the module that's providing the `Source`, `Sink`, and `Barrier`?

owen-mc commented on go/ql/src/Security/CWE-020/MissingRegexpAnchor.ql:
```diff
@@ -60,22 +60,22 @@ predicate isInterestingUnanchoredRegexpString(string re, string msg) {
-
-  predicate isSourceString(DataFlow::Node source, string msg) {
+module Config implements DataFlow::ConfigSig {
+  additional predicate isSourceString(DataFlow::Node source, string msg) {
```
The only difference I'm aware of is that if you want to make the module private then you have to move the predicate out of it.

mbg replied:
I suppose in any case it would be good to have consistency in which approach is used.

owen-mc commented on go/ql/test/library-tests/semmle/go/dataflow/InterProceduralDataFlow/Test.ql:
I wanted to make this use InlineFlowTest, or failing that InlineExpectationsTest, but I couldn't easily do it. I didn't want to throw away a lot of the information that is currently in the .expected file. But that's okay, it's okay to have tests which aren't InlineFlowTests or InlineExpectationsTests.

owen-mc commented on go/ql/lib/semmle/go/security/StringBreak.qll:
```diff
@@ -31,4 +33,21 @@ module StringBreak {
+    class FlowState = Quote;
+
+    predicate isSource(DataFlow::Node source, FlowState state) {
+      source instanceof Source and exists(state)
```
Just to make the warning that `state` isn't used go away. Maybe it's okay to have that warning.

mbg replied:
I have been trawling through the CodeQL docs and Slack a bit to see if there's a better way to do this, but couldn't find anything.

owen-mc commented on go/ql/test/query-tests/Security/CWE-190/AllocationSizeOverflow.expected:
It generally isn't obvious why. I know one cause is when there were more than one config in scope in the old version. I'm not sure how much value there is in tracking it down. I'm going to bet that in the corresponding PRs for java and C# they have similar changes and they haven't looked into what caused them. (Of course if there was a change in the results it would be different. It's just because the only change is in path nodes and edges.)

owen-mc commented on config/identical-files.json:
```diff
@@ -603,4 +602,4 @@
     "java/ql/lib/semmle/code/java/security/internal/EncryptionKeySizes.qll"
   ]
-}
+}
```
No

owen-mc commented on go/ql/test/library-tests/semmle/go/dataflow/VarArgsWithFunctionModels/Flows.ql:
```diff
@@ -18,69 +16,15 @@ class SummaryModelTest extends DataFlow::FunctionModel {
-    // or
-    // this.hasQualifiedName("github.com/nonexistent/test", "FunctionWithVarArgsOfStructsParameter")) and
-    // (inp.isParameter(0) and outp.isResult)
+    or
+    this.hasQualifiedName("github.com/nonexistent/test", "FunctionWithSliceOfStructsParameter") and
+    (inp.isParameter(0) and outp.isResult())
+    or
+    this.hasQualifiedName("github.com/nonexistent/test", "FunctionWithVarArgsOfStructsParameter") and
+    (inp.isParameter(0) and outp.isResult())
```
I added this comment in the test
```
	// The following tests all fail because FunctionModel doesn't interact with access paths
```
I believe that when I originally wrote the tests and realised this I just left the models commented out. This time I thought it was better to have the models there and the tests still failing. I don't know what's best. Maybe just removing those tests, and having a comment noting this particular shortcoming of `FunctionModel`?

owen-mc commented on go/ql/src/Security/CWE-089/StringBreak.expected:
Oops, this is from running the query like it was a test. It analyzes the examples from the qhelp. I'll delete it.

owen-mc commented on go/ql/lib/semmle/go/security/UnsafeUnzipSymlink.qll:
```diff
@@ -33,18 +35,31 @@ module UnsafeUnzipSymlink {
+    predicate isBarrier(DataFlow::Node node) { node instanceof EvalSymlinksInvalidator }
+  }
+
+  private module EvalSymlinksFlow = TaintTracking::Global<EvalSymlinksConfig>;
```
Are you saying that I should keep it public? I agree that that would be more backwards-compatible. But as this is a helper I think it should always have been private.

mbg replied:
It was the same comment as for the other file where the configuration was public and the `Flow` is private, so it is the same resolution here as well, which I think you have already done.

owen-mc commented on go/ql/src/experimental/CWE-74/DsnInjectionCustomizations.qll:
```diff
@@ -25,6 +28,20 @@ class DsnInjection extends TaintTracking::Configuration {
+private module DsnInjectionConfig implements DataFlow::ConfigSig {
+  predicate isSource(DataFlow::Node source) { source instanceof Source }
+
+  predicate isSink(DataFlow::Node sink) {
+    exists(Function f | f.hasQualifiedName("database/sql", "Open") |
+      sink = f.getACall().getArgument(1)
+    )
+  }
```
No. That line was added very recently, after I'd converted it, and I didn't spot it when I rebased. Well spotted!

owen-mc commented:
I believe I've addressed all comments except explaining the changes in test expectations.

owen-mc commented:
I've added explanations of test result changes in the relevant commit messages. I had to force-push to do this, so I rebased on `main` at the same time. There were no changes to the commits except the commit messages.

mbg commented on go/ql/lib/semmle/go/security/LogInjection.qll:
```diff
@@ -15,6 +15,8 @@ module LogInjection {
   import LogInjectionCustomizations::LogInjection
 
   /**
+   * DEPRECATED: Use `Flow` instead.
+   *
```
Can't comment on the relevant line because it wasn't changed as part of this PR, but `Flow` here doesn't have any documentation.

mbg commented on go/ql/lib/semmle/go/security/OpenUrlRedirect.qll:
```diff
@@ -63,4 +65,45 @@ module OpenUrlRedirect {
+    }
+  }
+
+  /** Tracks taint flow for reasoning about unvalidated URL redirections. */
```
```suggestion
  /** Tracks taint flow from unvalidated, untrusted data to URL redirections. */
```

mbg commented on go/ql/lib/semmle/go/security/ReflectedXss.qll:
```diff
@@ -35,4 +37,15 @@ module ReflectedXss {
+    predicate isBarrier(DataFlow::Node node) { node instanceof Sanitizer }
+  }
+
+  /** Tracks taint flow for reasoning about XSS. */
```
```suggestion
  /** Tracks taint flow from untrusted data to XSS attack vectors. */
```

mbg commented on go/ql/lib/semmle/go/security/RequestForgery.qll:
```diff
@@ -47,4 +49,24 @@ module RequestForgery {
+    }
+  }
+
+  /** Tracks taint flow for reasoning about request forgery. */
```
```suggestion
  /** Tracks taint flow from untrusted data to request forgery attack vectors. */
```

mbg commented on go/ql/lib/semmle/go/security/StringBreak.qll:
```diff
@@ -31,4 +33,25 @@ module StringBreak {
+  }
+
+  /**
+   * Tracks taint flow for reasoning about unsafe-quoting vulnerabilities,
+   * parameterized with the type of quote being tracked.
```
```suggestion
   * Tracks taint flow from untrusted data which may contain single or double quotes to uses where those quotes need to be escaped. 
   * The type of quote is accessible through the `Sink`.
```

mbg commented on go/ql/lib/semmle/go/security/UnsafeUnzipSymlink.qll:
```diff
@@ -14,9 +14,11 @@ module UnsafeUnzipSymlink {
   import UnsafeUnzipSymlinkCustomizations::UnsafeUnzipSymlink
 
   /**
+   * DEPRECATED: Use a copy of `EvalSymlinksConfig` and `EvalSymlinksFlow` instead.
```
```suggestion
   * DEPRECATED: Use copies of `EvalSymlinksConfig` and `EvalSymlinksFlow` instead.
```

mbg commented on go/ql/src/experimental/CWE-1004/CookieWithoutHttpOnly.ql:
```diff
@@ -78,9 +78,9 @@ predicate isGorillaSessionsCookieFlow(
           sink = sessionSave2 and
-          sessionSave2.getNode() = sessionSave.getNode()
+          source = options and
+          sessionSave.getNode() = sessionSave2.getNode()
```
Swap the two sides around for even fewer unnecessary changes!

mbg reviewed (approved):
Nice work investigating the reasons for the changes in the expected test output! Those comments are really useful, and more varied than I expected. 

I think this is probably good to merge as is. I have added a few minor bits of feedback for the new documentation, which might be nice to spend to improve if you have time, but otherwise aren't urgent changes.

Other comments:

- 855677c62514138f8eddafbfc0ded819b6204330 has meaningful changes to the `.expected` files, but no explanation. 
- b0302b6da09e1ec402b3effbeb5dd14a87dd9aa0 still has a stray `f` at the end of the commit message.

owen-mc commented on go/ql/lib/semmle/go/security/CleartextLogging.qll:
```diff
@@ -56,4 +58,43 @@ module CleartextLogging {
+  }
+
+  /**
+   * Tracks data flow for reasoning about clear-text logging of sensitive
+   * information, from `Source`s, which are sources of sensitive data, to
+   * `Sink`s, which is an abstract class representing all the places sensitive
+   * data may be stored in cleartext. Additional sources or sinks can be added
+   * by extending the relevant class.
```
This is copied from elsewhere in the file. I'm sure there are many improvements we could make to our documentation, but I suggest this could be done in a separate PR afterwards.

mbg reviewed (approved):
🚢
